engine/src/main/kotlin/world/gregs/voidps/engine/entity/character/player/skill/exp/Experience.kt

```diff
@@ -31,15 +32,29 @@ class Experience(
         events.emit(GrantExp(skill, previous, experience))
     }
 
-    fun add(skill: Skill, experience: Double) {
-        if (experience <= 0.0) {
+    fun add(skill: Skill, baseExperience: Double) {
+        if (baseExperience <= 0.0) {
             return
         }
+
+        val xpModSetting = Settings["world.dynamicXPMod", "off"].lowercase()
+        val playerMaxLevel = level(skill, get(skill))
+        val maxLevel = Level.MAX_LEVEL
+
+        val modifier = when (xpModSetting) {
+            "linear" -> interpolate(1.0, 5.0, playerMaxLevel.toDouble() / maxLevel)
+            "inverse" -> interpolate(5.0, 1.0, playerMaxLevel.toDouble() / maxLevel)
+            else -> 1.0
+        }
+
+        val experienceRate = Settings["world.experienceRate", DEFAULT_EXPERIENCE_RATE]
+        val adjustedExperience = baseExperience * modifier * experienceRate
+
```

Comment:
This is a bit too custom for my liking, 5.0 is very arbitrary, linear/inverse is redundant as you can just use negation and stacking this with the xp rate modifier feels odd.

It seems what you're really after here is an `experienceRateMin = 1.0` and `experienceRateMax = 1.0` where you can interpolate between those values. That should keep it more flexible and cleaner.